pkg/resolvers/k8s/watcher.go

```diff
@@ -39,63 +57,103 @@ func startNewWatcher(
 	watcherErrs prometheus.Counter,
 	watcherGotChanges prometheus.Counter,
 ) (*watcher, error) {
-	// NOTE(bplotka): Would love to have proper context from above but naming.Resolver does not allow that.
 	ctx, cancel := context.WithCancel(context.Background())
-
-	s, err := startNewStreamer(ctx, target, epClient)
-	if err != nil {
-		return nil, err
-	}
-
 	return &watcher{
-		logger:     logger,
-		ctx:        ctx,
-		cancel:     cancel,
-		target:     target,
-		streamer:   s,
-		addrsState: map[string]struct{}{},
-
+		logger:            logger,
+		ctx:               ctx,
+		cancel:            cancel,
+		target:            target,
+		epClient:          epClient,
+		addrsState:        map[string]struct{}{},
 		resolvedAddrs:     resolvedAddrs,
 		watcherErrs:       watcherErrs,
 		watcherGotChanges: watcherGotChanges,
+		streamer:          nil,
+		streamRetryBackoff: &backoff.Backoff{
+			Min:    50 * time.Millisecond,
+			Jitter: true,
+			Factor: 2,
+			Max:    2 * time.Second,
+		},
 	}, nil
 }
 
 // Close closes the watcher, cleaning up any open connections.
 func (w *watcher) Close() {
+	if w.streamer != nil {
+		w.streamer.Close()
+	}
+
 	w.cancel()
 }
 
+func (w *watcher) startNewStreamerWithRetry(ctx context.Context) (s *streamer) {
+	if ctx.Err() != nil {
+		return nil
+	}
+	var err error
+	for {
+		s, err = startNewStreamer(w.target, w.epClient)
+		if err == nil {
+			break
```

Comment:
nit:
```
if err == nil {
  w.streamRetryBackoff.Reset()
  return s
}
```
- don't used named returns, it's against improbable style guide (i think)
- you don't have to return anything after `for{}`.